src/lib/project-saver-hoc.jsx

```diff
@@ -118,25 +119,36 @@ const ProjectSaverHOC = function (WrappedComponent) {
          * @param {?object} requestParams - object of params to add to request body
          */
         storeProject (projectId, requestParams) {
-            return this.props.vm.saveProjectSb3()
-                .then(content => {
-                    const assetType = storage.AssetType.Project;
-                    const dataFormat = storage.DataFormat.SB3;
-                    const body = new FormData();
-                    body.append('sb3_file', content, 'sb3_file');
-                    if (requestParams) {
-                        for (const key in requestParams) {
-                            body.append(key, requestParams[key]);
-                        }
-                    }
-                    // when id is undefined or null, storage.store as we have
-                    // configured it will create a new project with id
-                    return storage.store(
-                        assetType,
-                        dataFormat,
-                        body,
-                        projectId
-                    );
+            requestParams = requestParams || {};
+            return Promise.all(this.props.vm.assets
+                .filter(asset => !asset.clean)
+                .map(
+                    asset => storage.store(
+                        asset.assetType,
+                        asset.dataFormat,
+                        asset.data,
+                        asset.assetId
+                    ).then(
+                        () => (asset.clean = true)
+                    )
+                )
+            ).then(() => {
+                const body = new FormData();
+                const sb3Json = new Blob([this.props.vm.toJSON()], {type: 'application/json'});
+                body.append('sb3_file', sb3Json, 'sb3_file');
+                for (const key in requestParams) {
+                    if (requestParams.hasOwnProperty(key)) body.append(key, requestParams[key]);
+                }
+                return storage.store(
+                    storage.AssetType.Project,
+                    storage.DataFormat.JSON,
+                    body,
+                    projectId
+                );
+            })
+                .catch(err => {
```

Comment:
I'm not sure, but I think it's a mistake to intercept this error here, rather than pass it up to the catch() calls that encase the calls to storeProject().

Reply:
This is kind of here as a placeholder for what we want to do in https://github.com/LLK/scratch-gui/issues/3723. I think we will probably want to handle all of that here, close to where the actual saving is happening.